README.md
# Get Tension

For people who love Tennis and Badminton.
This is an application for estimating racket string tension with AI.
You can get racket string tension from the sound by hitting your racket face.
AI learns sounds just finished stringing and predicts current racket string tension.
It works on Mac, Windows, Linux.
It is written in Python.
By default, training data is my stringing data. If you are a stringer, you can use your own stringing data and get your own tension, too.

![python powered](https://raw.githubusercontent.com/sugi-juku/get-tension/master/python-powered-w-200x80.png)

## Description

My stringing shop was closed about 2 months for COVID-19 in Spring 2020.
So I decided to develop an application getting racket string tension from sound.

There is also a method of measuring Dynamic Tension (DT) with a dedicated measuring machine.
But when you order stringing, you can only specify the tension of the stringing machine.
So I want to know what the tension of the stringing machine is now.

The work is as follows.

1. I assumed that the racket string sound just finished stringing showed the machine tension.
1. I considered the elements necessary for learning. For example, the sound hitting racket face, string gauge, string material, stringing pattern, racket face size, etc.
1. I recorded racket string sounds just finished stringing and gathered training data.
1. I got fundamental frequency from sounds by FFT.
1. I tried to let AI learn training data set by scikit-learn Linear Regression.
1. I tried to design good classes.
1. CUI version worked on macOS Python 2.7.16.
1. I tried to add GUI with tkinter.
1. GUI version worked on macOS Python 2.7.16.
1. It worked on macOS Python 3.7.7.
1. Apple says "Use of Python 2.7 isn\'t recommended as this version is include in macOS for compatibility with legacy software.". So I decided to use Python3.
1. It worked on Ubuntu Python 3.7.5.
1. It worked on Windows 10 Home Python 3.8.3.

I\'m both a stringer and a programmer.
I used to develop web applications often with PHP at several major IT companies.
This is my first application with Python.
Now, I love Python, too.

## Requirement

Maybe it works on Python 3.6 or higher.
It works on Mac, Windows, Linux.
### Tested environment

* macOS Intel 10.14.6 - 13.4.1 (My development environment)

    * python 3.7.7 - 3.9.13

* Ubuntu 19.10 - 20.04 - 22.04

    * pytyon 3.7.5 - 3.10.6

* Windows 10 Home 1909

    * python 3.8.3

## Installation

Download from https://github.com/sugi-juku/get-tension and deploy your PC.

For example.

```
git clone https://github.com/sugi-juku/get-tension.git
```

### Python and Python modules

#### macOS

Install Python3.X for macOS from https://www.python.org/downloads/ .

Then, Install python modules.

```
./install_python_modules_mac.sh
```

#### Windows

Install Python3.X for Windows from https://www.python.org/downloads/ .
If your PC is x86-64, download and install  ```Windows x86-64 executable installer``` .

Then install python modules. Double click ```install_python_modules_win.bat``` .

#### Ubuntu

Install python and python modules.

```
./install_python_modules_linux.sh
```

## Usage

If you use Windows, you can run ```*.py``` with a double click.

- GUI version

    It is easy to use.

    ```
    python3 get_tension_app.py
    ```

    ![app top image](https://raw.githubusercontent.com/sugi-juku/get-tension/master/app_top.png)

    If it doesn't work well, try to increase the sensitivity of the microphone.

    If you use macOS, you can run ```get_tension.command``` with a double click.

- CUI version

    If you are a stringer, try CUI version.
    Otherwise you don\'t have to use CUI version.

    If you use Windows, replace ```python3``` with ```py``` .

    - get_tension_data_append.py

        If you have just finished stringing, use this to get racket string tension.
        Data is appended to training data set.
        For example.

        ```
        python3 get_tension_data_append.py T_45-43_GOMS16-GOAKP16_98_16-19
        python3 get_tension_data_append.py B_25_YOBG66UM_56
        ```
    - get_tension_data_noappend.py

        If you want to get current racket string tension, use this.
        For example.

        ```
        python3 get_tension_data_noappend.py T_45-43_GOMS16-GOAKP16_98_16-19
        python3 get_tension_data_noappend.py B_25_YOBG66UM
        ```

        - Argument: Tennis

            ```
            Argument = T_{MainTension}-{CrossTension}_{MainString}-{CrossString}_{FaceSize}_{MainStringNumber}-{CrossStringNumber}

            if MainTension == CrossTension:
                Argument = T_{MainTension}_{MainString}-{CrossString}_{FaceSize}_{MainStringNumber}-{CrossStringNumber}

            if MainString == CrossString:
                Argument = T_{MainTension}-{CrossTension}_{MainString}_{FaceSize}_{MainStringNumber}-{CrossStringNumber}
            ```

        - Argument: Badminton

            ```
            Argument = B_{MainTension}-{CrossTension}_{MainString}-{CrossString}_{FaceSize}

            if MainTension == CrossTension:
                Argument = B_{MainTension}_{MainString}-{CrossString}_{FaceSize}
            
            if MainString == CrossString:
                Argument = B_{MainTension}-{CrossTension}_{MainString}_{FaceSize}
            ```

    - get_tension_from_wav.py

        You can get racket string tension from wavfile.
        {wavfilename} is filename with a relative path from get_tension_from_wav.py.

        ```
        python3 get_tension_from_wav.py {wavfilename}
        ```

    - plt_data.py

        You can see graphs of training data set.
        This includes fit_lr.py.

        ```
        python3 plt_data.py
        ```

    - plt_tension_change.py

        You can see a graph of tension change from wavfile in target directory.
        For example.

        ```
        python3 plt_tension_change.py tmp/sample
        ```

        ![GOSEN AK PRO 16](https://raw.githubusercontent.com/sugi-juku/get-tension/master/GOAKP16.png)

    - fit_lr.py

        Linear model is fitted by training data set.
        It is good to run this regularly using cron.

    - remake_stdata.py

        You can remake training data set from wavfiles.
        Normally you don\'t have to use this.


## Author

* Yukihiko Sugimura
* E-Mail : sugimura_juku@yahoo.co.jp
* https://sugi-juku.com
* https://tennis.sugi-juku.com

## Licence

Copyright (c) 2020 Yukihiko Sugimura

Released under the [MIT license](https://opensource.org/licenses/mit-license.php).


changelog.md
# Version 1.8

Released 2023-07-15

- Modified about f0 search method
- Added some strings
- Number of training data: Tennis=1042, Badminton=279

# Version 1.7

Released 2022-03-05

- Added correcting tension for each string when hybrid
- Added some strings
- Modified documents

# Version 1.6

Released 2021-11-12

- Some bugfixes
- Added correcting tension for each string when not hybrid
- Added "Back to top" button
- Added some strings
- Modified about Young's modulus value

# Version 1.5

Released 2021-05-01

- Some bugfixes
- Added Nylon multi Young's modulus
- Changed badminton stringing machine GOSEN GM20-EX
- Changed badminton tension calculating logic tension = main_tension
- Modified to check measurement tension value before appending
- Added Wilson Badminton strings

# Version 1.4

Released 2020-08-21

- Some bugfixes
- Added Badminton face size to Training Data

# Version 1.3

Released 2020-08-10

- Some bugfixes
- Added README.ja.md
- Modified about FFT search range
- Changed to use class StringLr01
- Added String Young's Modulus to Training Data

# Version 1.2

Released 2020-07-03

- Some bugfixes
- Worked on Windows 10 Home Python 3.8.3
- Worked on Ubuntu 20.04 Python 3.8.2
- Worked on Mac OS X 10.14.6 Python 3.7.8
- Replaced sound modules pyaudio and wav with sounddevice and soundfile
- Changed \_\_file\_\_ to sys.argv[0]
- Modified installer
- Added get_mean_absolute_error()
- Added get_r2_score()

# Version 1.1

Released 2020-06-24

- Some bugfixes
- Worked on Ubuntu 19.10 Python 3.7.5
- Decided to use Python3
- Changed to use class StringLr00
- Added get_mean_squared_error()

# Version 1.0

Released 2020-06-17

- First public release
- Tested on Mac OS X 10.14.6
- Worked on Python2.7.16 and Python3.7.7


get_tension_app.py
#!/usr/bin/env python3
# coding: utf-8

import os
import sys

# Python2
# import Tkinter as tk
# import tkMessageBox as messagebox
# import ttk

# Python3
import tkinter as tk
from tkinter import messagebox
import tkinter.ttk as ttk

os.chdir(os.path.dirname(os.path.abspath(sys.argv[0])))

import string_def as sde
import string_rec as sre
import string_data as sda
import string_lr as slr


def run_click():
    # run_btn.grid_forget()
    # data_exp_label_text.set("Recording...")
    messagebox.showinfo("Rcording", "After ckicking [OK] button,\nplease hit your racket face near PC.")

    sdef = sde.StringDef(stype)
    main_s = ""
    cross_s = ""
    if main_s_comb.get() != "":
        main_s = sdef.get_key_from_name(main_s_comb.get())
    if cross_s_comb.get() != "":
        cross_s = sdef.get_key_from_name(cross_s_comb.get())

    args = ""
    if stype == "B":
        args = sda.get_argstr(stype, main_t_comb.get(), cross_t_comb.get(), main_s, cross_s, size_comb.get(), "", "")
    if stype == "T":
        args = sda.get_argstr(stype, main_t_comb.get(), cross_t_comb.get(), main_s, cross_s, size_comb.get(), main_n_comb.get(), cross_n_comb.get())

    csv_append = 0
    if data_comb.get() == "Yes":
        csv_append = 1

    if csv_append == 0:
        srec = sre.StringRec(args, "tmp/")
    if csv_append == 1:
        srec = sre.StringRec(args)

    # data_exp_label_text.set("Calculating...")

    tension = []
    for filename in srec.get_file_list():
        sdata = sda.StringData()
        sdata.make_data(filename)
        stglr = slr.StringLr01(stype)
        xlist = stglr.get_lrdata_xlist(sdata)
        mst = sdata.get_main_string()
        cst = sdata.get_cross_string()
        tension.append(stglr.get_lrcal_tension_cor(xlist, mst, cst))

    tension_error = 0.0
    for i, val in enumerate(tension):
        tension_error += abs(tension[0]-val)
    tension_error = tension_error / (len(tension) - 1)
    print("MAError = " + str(tension_error))

    rstr = ""
    if tension_error > sda.tension_error_val[stype]:
        for filename in srec.get_file_list():
            os.remove(filename)
            print("Removed: " + filename)
        rstr = "Measurement error.\nPlease try again."
    else:
        if csv_append == 1:
            for filename in srec.get_file_list():
                sdata = sda.StringData()
                sdata.make_data(filename, csv_append)
        total = 0.0
        avg = 0.0
        for i, val in enumerate(tension):
            rstr += str(i+1) + ": " + str(round(val, 2)) + " lbs\n"
            total += float(val)
        avg = total/len(tension)
        rstr += "Average: " + str(round(avg, 2)) + " lbs"

    srec.init_file_list()

    messagebox.showinfo("Results", rstr)
    print(rstr)

    main_win.focus_force()
    # main_win.destroy()

def make_win():
    global main_win
    main_win = tk.Tk()
    win_w = main_win.winfo_screenwidth()
    win_h = main_win.winfo_screenheight()
    app_w = 530
    app_h = 285
    if stype == "B":
        app_h = 230
    main_win.geometry(get_center_geometry_str(win_w, win_h, app_w, app_h))
    main_win.title("Stringing DATA")

    main_frm = ttk.Frame(main_win)
    main_frm.grid(column=0, row=0, sticky=tk.NSEW, padx=5, pady=10)

    sdef = sde.StringDef(stype)
    if stype == "T":
        t_list = list(range(0,81))
        t_current = 50
    if stype == "B":
        t_list = list(range(0,41))
        t_current = 20

    main_t_label = ttk.Label(main_frm, text="Main Tension (lbs)")
    global main_t_comb
    main_t_comb = ttk.Combobox(main_frm, values=t_list, state="readonly", width=5)
    main_t_comb.current(t_current)
    cross_t_label = ttk.Label(main_frm, text="Cross Tension (lbs)")
    global cross_t_comb
    cross_t_comb = ttk.Combobox(main_frm, values=t_list, state="readonly", width=5)
    main_s_label = ttk.Label(main_frm, text="Main String")
    global main_s_comb
    main_s_comb = ttk.Combobox(main_frm, values=sdef.get_name_list(), state="readonly", width=30)
    main_s_comb.current(0)
    cross_s_label = ttk.Label(main_frm, text="Cross String")
    global cross_s_comb
    cross_s_comb = ttk.Combobox(main_frm, values=sdef.get_name_list(), state="readonly", width=30)

    if stype == "T":
        s_list = list(range(70,141))
        s_current = 30
    if stype == "B":
        s_list = list(range(50,61))
        s_current = 6

    size_label = ttk.Label(main_frm, text="Face Size (in2)")
    global size_comb
    size_comb = ttk.Combobox(main_frm, values=s_list, state="readonly", width=5)
    size_comb.current(s_current)

    if stype == "T":
        pat_list = list(range(0,31))
        main_n_label = ttk.Label(main_frm, text="Main Number")
        global main_n_comb
        main_n_comb = ttk.Combobox(main_frm, values=pat_list, state="readonly", width=5)
        main_n_comb.current(16)
        cross_n_label = ttk.Label(main_frm, text="Cross Number")
        global cross_n_comb
        cross_n_comb = ttk.Combobox(main_frm, values=pat_list, state="readonly", width=5)
        cross_n_comb.current(19)

    data_label = ttk.Label(main_frm, text="Append DATA")
    global data_exp_label
    global data_exp_label_text
    data_exp_label_text = tk.StringVar()
    data_exp_label_text.set("If you have just finished stringing, select Append DATA = Yes.")
    data_exp_label = ttk.Label(main_frm, textvariable=data_exp_label_text)
    global data_comb
    data_comb = ttk.Combobox(main_frm, values=["Yes", "No"], state="readonly", width=5)
    data_comb.current(1)
    global run_btn
    run_btn = ttk.Button(main_frm, text="RUN", command=run_click)
    global top_btn
    top_btn = ttk.Button(main_frm, text="Back to top", command=top_click)

    main_t_label.grid(column=0, row=0)
    main_t_comb.grid(column=1, row=0, sticky=tk.W, padx=5)
    cross_t_label.grid(column=0, row=1)
    cross_t_comb.grid(column=1, row=1, sticky=tk.W, padx=5)
    main_s_label.grid(column=0, row=2)
    main_s_comb.grid(column=1, row=2, sticky=tk.W, padx=5)
    cross_s_label.grid(column=0, row=3)
    cross_s_comb.grid(column=1, row=3, sticky=tk.W, padx=5)
    size_label.grid(column=0, row=4)
    size_comb.grid(column=1, row=4, sticky=tk.W, padx=5)
    if stype == "T":
        main_n_label.grid(column=0, row=5)
        main_n_comb.grid(column=1, row=5, sticky=tk.W, padx=5)
        cross_n_label.grid(column=0, row=6)
        cross_n_comb.grid(column=1, row=6, sticky=tk.W, padx=5)
        data_label.grid(column=0, row=7)
        data_comb.grid(column=1, row=7, sticky=tk.W, padx=5)
        data_exp_label.grid(column=1, row=8)
        top_btn.grid(column=0, row=9)
        run_btn.grid(column=1, row=9)
    else:
        data_label.grid(column=0, row=5)
        data_comb.grid(column=1, row=5, sticky=tk.W, padx=5)
        data_exp_label.grid(column=1, row=6)
        top_btn.grid(column=0, row=7)
        run_btn.grid(column=1, row=7)

    main_win.columnconfigure(0, weight=1)
    main_win.rowconfigure(0, weight=1)
    main_win.columnconfigure(0, weight=1)

    main_win.mainloop()

def top_click():
    main_win.destroy()
    make_top()

def stype_click_t():
    global stype
    stype = "T"
    main_win.destroy()
    make_win()

def stype_click_b():
    global stype
    stype = "B"
    main_win.destroy()
    make_win()

def get_center_geometry_str(win_w, win_h, app_w, app_h):
    center_w = int((win_w-app_w)/2)
    center_h = int((win_h-app_h)/2)
    return str(app_w)+"x"+str(app_h)+"+"+str(center_w)+"+"+str(center_h)

def make_top():
    global main_win
    main_win = tk.Tk()

    win_w = main_win.winfo_screenwidth()
    win_h = main_win.winfo_screenheight()
    app_w = 300
    app_h = 330
    main_win.geometry(get_center_geometry_str(win_w, win_h, app_w, app_h))

    main_win.title("Get Tension")
    main_win.configure(bg="white")
    logo = tk.PhotoImage(file="logo_270.gif")
    logo_label = tk.Label(main_win, image=logo)
    py_logo = tk.PhotoImage(file="python-powered-w-200x80.gif")
    py_logo_label = tk.Label(main_win, image=py_logo)
    label = tk.Label(text="Please select.")
    copy_label = tk.Label(text="Copyright (C) 2020 Yukihiko Sugimura.")
    t_btn = ttk.Button(main_win, text="Tennis", command=stype_click_t)
    b_btn = ttk.Button(main_win, text="Badminton", command=stype_click_b)

    logo_label.pack()
    label.pack()
    t_btn.pack(fill="x")
    b_btn.pack(fill="x")
    py_logo_label.pack()
    copy_label.pack()

    main_win.mainloop()


make_top()
